LanguageFeatures/Records/type_inference_A02_t01.dart

```diff
@@ -2,9 +2,28 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-/// @assertion Type inference and promotion flows through records in much the 
-/// same way it does for instances of generic classes (which are covariant in 
-/// Dart just like record fields are) and collection literals.
+/// @assertion Given a type schema K and a record expression E of the general
+/// form (e1, ..., en, d1 : e{n+1}, ..., dm : e{n+m}) inference proceeds as
+/// follows.
+///
+/// If K is a record type schema of the form
+/// (K1, ..., Kn, {d1 : K{n+1}, ...., dm : K{n+m}}) then:
+///
+/// - Each ei is inferred with context type schema Ki to have type Si
+///   - Let Ri be the greatest closure of Ki
+///   - If Si is a subtype of Ri then let Ti be Si
+///   - Otherwise, if Si is dynamic, then we insert an implicit cast on ei to
+///     Ri, and let Ti be Ri
+///   - Otherwise, if Si is coercible to Ri (via some sequence of call method
+///     tearoff or implicit generic instantiation coercions), then we insert the
+///     appropriate implicit coercion(s) on ei. Let Ti be the type of the
+///     resulting coerced value (which must be a subtype of Ri, possibly proper).
+///   - Otherwise, it is a static error.
+///  - The type of E is (T1, ..., Tn, {d1 : T{n+1}, ...., dm : T{n+m}})
+/// If K is any other type schema:
+///
+/// - Each ei is inferred with context type schema _ to have type Ti
+/// - The type of E is (T1, ..., Tn, {d1 : T{n+1}, ...., dm : T{n+m}})
 ///
 /// @description Checks horizontal inference for records.
```

Comment:
I don't think horizontal inference is applicable to record literals (at least, I don't remember hearing anything about it). But it sounds like an interesting additional feature, so it's great that we have it on the table! I'm creating a new language repo issue about this question.

Next, I can see that this test is actually about a situation involving a generic function and an invocation thereof, and the new thing is that some types involved in the associated horizontal inference for that function invocation are record types. Very good, we need that!

As far as I can see, we have the stages `{b, c, d}` and `{a}`, and `T` and `U` are as stated. So that looks good!

See https://github.com/dart-lang/language/issues/2578 as well, we may need some additional tests (possibly that it _doesn't_ work, if the language and implementation teams think it should not be supported, or that it should wait).